clang/lib/CIR/CodeGen/CIRGenExprConst.cpp

```diff
@@ -1049,16 +1049,39 @@ class ConstantLValueEmitter
 
   bool hasNonZeroOffset() const { return !Value.getLValueOffset().isZero(); }
 
-  /// Return the value offset.
-  mlir::Attribute getOffset() { llvm_unreachable("NYI"); }
-
   // TODO(cir): create a proper interface to absctract CIR constant values.
 
+  /// Return GEP-like value offset
+  void getOffset(mlir::Type Ty,
+                 llvm::SmallVectorImpl<mlir::Attribute> &Indices) {
+    auto Offset = Value.getLValueOffset().getQuantity();
+    CIRDataLayout Layout(CGM.getModule());
+    SmallVector<int64_t, 3> Idx;
+    CGM.getBuilder().computeGlobalViewIndicesFromFlatOffset(Offset, Ty, Layout,
+                                                            Idx);
+
+    for (auto I : Idx) {
+      auto Attr = mlir::cir::IntAttr::get(CGM.getBuilder().getSInt64Ty(), I);
+      Indices.push_back(Attr);
+    }
+  }
+
   /// Apply the value offset to the given constant.
   ConstantLValue applyOffset(ConstantLValue &C) {
     if (!hasNonZeroOffset())
       return C;
 
+    if (auto Attr = C.Value.dyn_cast<mlir::Attribute>()) {
+      auto GV = cast<mlir::cir::GlobalViewAttr>(Attr);
+      assert(!GV.getIndices());
+
+      llvm::SmallVector<mlir::Attribute, 3> Indices;
+      getOffset(GV.getType(), Indices);
+
+      return mlir::cir::GlobalViewAttr::get(
+          GV.getType(), GV.getSymbol(), CGM.getBuilder().getArrayAttr(Indices));
```

Comment:
Can you keep the old `getOffset` signature and make it return `CGM.getBuilder().getArrayAttr(Indices)` directly?

Reply:
Unfortunately we need to pass the type, to be able to generate GEP indices...

Reply:
It's fine to pass the type, but should return `mlir::Attribute` or `mlir::ArrayAttr`.

Reply:
Ah, right, I missed that part. Fixed now.